### ISSUE TITLE
Lua 5.4 <const> and <close> variable attribute in LuaJIT?

### ISSUE THREAD

**Issue by MikuAuahDark** [LuaJIT/LuaJIT#1013]
**Note**: While this can be treated as feature request, I don't want this thing to be solely a feature request such as "Hey Mike please add this feature thanks". I'd like to know about other opinion about this (especially Mike) and the technical challenges of implementing this, if it's even possible. So perhaps, an RFC? Discusison?

-----

There have been related issue/PRs about adding/backporting Lua 5.4 feature (#525, #590, #926, #927, #929, #938, and probably more that I haven't catch). However I don't think LuaJIT needs all of those Lua 5.4 changes.

In my opinion, the only Lua 5.4 change that is good enough withut changing LuaJIT as a whole is the `<const>` and `<close>` variable attributes. The `<const>` variable attribute ensure that you can't assign to same variable twice. However, I think `<close>` is a game changer.

The `<close>`, or to-be-closed variables is also a `<const>` variable, but with additional feature: When the variable goes out-of-scope (by using `break`, `goto`, `return` or **even Lua error**), the metamethod `__close` is called to the underlying value with 1st parameter is the object, and the 2nd parameter is the error object (if any). I think the technical difficulties on implementing this is when an error occurs as the `__close` metamethod must be called.

I think people is more interested with the to-be-closed attribute, as this can simplify certain things. Such use-case includes:

1. File operation
```lua
function doSomething()
    local f <close> = io.open("file.txt", "rb")
    doSomethingThatMayError(f)
end
```

In that example, `f:close()` will be automatically called when the function ends. Even if `doSomethingThatMayThrowError` throws error, `f:close()` is guaranteed to be called. Lua 5.4 has built-in integration of `__close` metamethod to `file*` objects.

2. Critical section
```lua
-- This is SDL example, but can be adapted to other APIs too
local ffi = require("ffi")
local SDL2 = ffi.load(ffi.os == "Windows" and "SDL2" or "libSDL2-2.0.so.0")
ffi.cdef [[
typedef struct SDL_mutex SDL_mutex;

SDL_mutex* SDL_CreateMutex(void);
int SDL_LockMutex(SDL_mutex * mutex);
int SDL_UnlockMutex(SDL_mutex * mutex)
void SDL_DestroyMutex(SDL_mutex * mutex);
]]

local lockguard_mt = {
    __close = function(mutex, err)
        SDL2.SDL_UnlockMutex(mutex.mutex)
    end,
}

function lockguard(mutex)
    assert(SDL2.SDL_LockMutex(mutex) == 0, "mutex acquire error")
    return setmetatable({mutex = mutex}, lockguard_mt)
end

local doSomethingMutex = SDL2.SDL_CreateMutex()
-- Pass `doSomethingMutex` to other thread

-- Meanwhile in other thread
local doSomethingMutex = getSharedMutexFromSomewhere()

function doSomething(arraybuffer)
    -- Only allow one thread to execute this code so only 1 thread can access arraybuffer.
    local lock <close> = lockguard(doSomethingMutex)
    -- do something with arraybuffer without worrying about race condition
    if certain_condition_not_satisfied then
        error("something is wrong")
    end
end
```
In this example, only 1 thread can run all code in `doSomething`. The mutex is acquired by the `lock` variable through `lockguard` function that returns table with `__close` metamethod. If the `lockguard` throws error, the to-be-closed variable won't call `__close` metamethod. If code around the function throws error then the mutex is automatically released, thus it's _safe_. If someone knows C++, this example is similar to `std::lock_guard`.

But I'm sure there's certain edge case which needs to be handled, such as:
1. **What if I return the said to-be-closed variables?** Then the `__close` metamethod must be called, so the returned object will be "closed" (depending on the purpose).
2. **How will `<const>` and `<close>` affect GC?** _I don't think_ it affects GC significantly (if any). A "const" variable is only meant for variable assignment checking and they still needs to be tracked by the garbage collector because they're not finalized yet. But it can be used as a hint to garbage collect the objects early in the future (as separate feature). For "closed" variable, the `__close` and `__gc` metamethod serve 2 different things. A "closed" object still allocate some memory in it.
3. **Would bytecode compatibility guaranteed?** Unfortunately, I don't think bytecode compatibility is guaranteed. Although I'm not expert on how bytecodes are defined internally, so I can't be certain.

So, I'd like to hear other opinion about this and for LuaJIT experts, the technical difficulties on implementing this feature. I'm sure it sounds easy but hard in reality so I don't want to brag about it too much. If this gets implemented, then great. If it doesn't, then it's fine.

**Comment by XmiliaH** [comment]
I think this would not be trivial to implement. First it requires a side structure to track the to-be-closed values. In PUC Lua they are an upvalue and the tracking is done on the stack as they had some bytes left per stack slot. In LuaJIT there is no space for this on the stack except if to-be-closed values would allocate two stack slots. Furthermore, in PUC Lua all to-be-closed values are upvalues and are closed with the BC_CLOSE bytecode. If it would be implemented the same way in LuaJIT is would not allow for JITing these constructs as the equivilant bytecode is NYI, and even if it would be implemented it would likely not allow for the to-be-closed value to be sunken as it would need to be tracked.

Alltogether, I think this could be implemented by allocating a second stack slot for to-be-closed values which saves the link to the next to-be-closed stack slot but make it that to-be-closed values do not need to be upvalues. This allows the BC_CLOSE to check if it only closes to-be-closed values and no upvalues and allows for JITing in such cases.

**Comment by MikePall** [comment]
No way. That feature is so misdesigned, it hurts.

Please look at the change history of Lua 5.4. After three years they are still playing the whack-a-mole game, trying to fix the collateral damage this caused everywhere in the VM (scopes, frames, coroutines, tailcalls, generic for loop, Lua/C API and so on). E.g. formerly `lua_pop()` was a simple pointer decrement (`L->top--`). Now, try to figure out what the current code expands to and start to cry.

First, the benefit for a new syntactic and semantic construct to provide locally scoped deterministic finalization in a garbage collected language is low.
This mainly helps for toy programs or for initializer functions that set up multiple resources. The latter are few and far between -- we coped just fine without all of that extravaganza.
One still has to remember to explicitly close files, sockets and so on. In non-toy programs this usually happens in a completely different place, i.e. not in the same local scope. That feature doesn't help you at all with this.

It helps to consider the equivalent feature in non-garbage collected languages (like C++) is 99% of the time used to deallocate memory (and not resources). Which is a complete non-issue in Lua.

Second, there's already a perfectly working mechanism in Lua called a protected frame aka `pcall()`. In LuaJIT this is even zero-cost. The correct approach would have been to reuse that mechanism to avoid creating new intra-frame finalization semantics. And an explicit block syntax wouldn't hurt, too. Because that Lua 5.4 feature right now creates an implicit semantic-block-but-not-really-a-block, which is not only hard to understand but is the root cause for most of its technical issues.

tl;dr: The cost-benefit ratio is very poor.

**Comment by MikuAuahDark** [comment]
I see. The kind of answers from XmiliaH and Mike is what I'm looking for and thank you for both of your explanation on their technical difficulties and reason it should**n't** be in LuaJIT.

**Comment by mcepl** [comment]
> Second, there's already a perfectly working mechanism in Lua called a protected frame aka `pcall()`.

This comment somehow confuses me. Are you objecting against whole idea of automatic closing or just against the Lua PUC implementation? Related: what do you think about context managers in Python? We, in the Python universe, generally like them very much and they help a lot to prevent leaked resources, unclosed sockets, connections, etc.

**Comment by GitSparTV** [comment]
The downside is you need to wrap entire code written after initialized resource in pcall in order to destroy the resource on error.
If there were no errors, close variables would be easier: just place defer code on every return.

Pcall for every resource makes code cluttered same way as try catch in C++
We have bunch of cases when malloced memory must be freed explicitly
